keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,380 @@
+	if err != nil {
+		klog.V(3).Infof("error reading persistent volume %q: %v", volume.Name, err)
+		return "", nil
+	}
+
+	if newVolume.GetDeletionTimestamp() != nil {//<==========================DeletionTimestamp set since the PV was deleted first.
+		klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+		return "", nil
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+11. Meanwhile, the external-provisioner checks if there is a `deletionTimestamp` on the PV, if so, it assumes that its
+in a transitory state and returns false for the `shouldDelete` check.
+```go
+// shouldDelete returns whether a volume should have its backing volume
+// deleted, i.e. whether a Delete is "desired"
+func (ctrl *ProvisionController) shouldDelete(ctx context.Context, volume *v1.PersistentVolume) bool {
+	if deletionGuard, ok := ctrl.provisioner.(DeletionGuard); ok {
+		if !deletionGuard.ShouldDelete(ctx, volume) {
+			return false
+		}
+	}
+
+	if ctrl.addFinalizer && !ctrl.checkFinalizer(volume, finalizerPV) && volume.ObjectMeta.DeletionTimestamp != nil {
+		return false
+	} else if volume.ObjectMeta.DeletionTimestamp != nil { //<========== DeletionTimestamp is set on the PV since it's deleted first.
+		return false
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+
+The main approach in fixing the issue involves introducing new finalizer `kubernetes.io/pv-deletion-protection` on the Persistent Volume when the controller sees it. Adding the finalizer prevents the PV to be removed from the API server. The finalizer will be removed only after the physical volume on the storage system is deleted.
+
+When the PVC is deleted, the PV is moved into `Released` state and following checks are made:
+
+1. Plugin:
+If the volume has the finalizer `kubernetes.io/pv-deletion-protection`, then, `DeletionTimestamp` checks can be ignored.
+
+3. CSI driver
+
+If when processing the PV update it is observed that it has `kubernetes.io/pv-deletion-protection` finalizer and
+`DeletionTimestamp` set, then the volume deletion request is forwarded to the driver(provided other pre-defined 
+conditions are met).
+
+
+### Risks and Mitigations
+
+The primary risk is volume deletion to a user that expect the current behavior, i.e, they do not expect the volume to be
+deleted from the underlying storage infrastructure when for a bound PV-PVC pair, the PV is deleted followed by deleting
+the PVC. This can be mitigated by initially introducing the fix behind a feature gate, followed by explicitly deprecating
+the current behavior.
+
+## Design Details
+
+A feature gate named `HonorPVReclaimPolicy` will be introduced for both `kube-controller-manager` and `external-provisioner`.
```

Comment:
*(Resuming our conversation on slack, where we were discussing if a feature gate or cmdline argument should be used).*

I've been thinking about it, having a feature gate looks better to me. We want all CSI drivers to have the same behavior and not some drivers protected and some not. In ideal case, all drivers will update their external-provisioner to the version where the protection is GA eventually and all of them will be the same. With cmdline option, some drivers could opt in, other drivers could opt out and for cluster admin / user it would be confusing which PVs are protected and which are not.

Reply:
Sure, will keep it a feature gate.